src/lib.rs

```diff
@@ -696,4 +708,72 @@ mod tests {
         // don't panic in drop impl.
         core.has_dropped_slots = true;
     }
+
```

Comment:
it would be nice to also add a `loom` model exercising what happens when a buffer fills up due to _concurrent_ pushes from multiple threads? we could do something where we spawn multiple threads and have each one try to push in a loop until the buffer is full, which would check that all of those threads eventually complete. 

we could also exercise slot skipping by adding a thread that calls `pop_ref` and either `mem::forget`s the guards or stuffs them someplace to hang onto them.

Reply:
@hawkw I need your advise here, I added a simple loom test but I find it difficult to construct a good test with read/writes under loom.  Imagine we have a thread that reads from the buffer (for example, exactly three times) and we have two threads that write to the buffer until three elements are read and the buffer is full, under loom we will fail because we will reach max iterations in cases in which we attempt to write a lot and doesn't read, making us spin in "buffer is full" and not progress